lib/praxis/mapper/resources/typed_methods.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+module Praxis
+  module Mapper
+    module Resources
+      # Error raised when trying to call a typed method and the validation of arguments fails
+      class IncompatibleTypeForMethodArguments < ::StandardError
+        attr_reader :errors, :method
+
+        def initialize(errors:, method:, klass:)
+          @errors = errors
+          super "Error validating/coercing arguments for call to method #{method} of class #{klass}:\n#{errors}"
+        end
+      end
+
+      module TypedMethods
+        extend ::ActiveSupport::Concern
+
+        included do
+          include Praxis::Mapper::Resources::Callbacks
+
+          class << self
+            attr_reader :signatures
+
+            def _finalize!
+              if @signatures
+                # Build the around callbacks for coercing the params for the methods with types defined
+                # Also, this needs to be before, so that we hit the coercion code before any other around callback
+                const_set(:MethodSignatures, Module.new)
+                @signatures.each do |method, type|
+                  # Also add a constant pointing to the signature type inside Signatures (and substitute ! for Bang, as that's not allowed in a constant)
+                  # This helps with debugging, as we won't get anonymous struct classes, but we'll see these better names
+                  cleaned_name = method.to_s.gsub(/!/, '_bang')
+                  self::MethodSignatures.const_set(cleaned_name.camelize.to_sym, type)
+                  coerce_params_for method, type
+                end
+              end
+
+              super
+            end
+
+            def signature(method, &block)
```

Comment:
Who needs Sorbet or RBS when you've got Praxis? 😄

Reply:
yeah, and now, we even have the splatting support for args...check it out! :)